### PR TITLE
reproducible builds: remove CFLAGS from :version in non-debug builds

### PR DESCRIPTION
**Pull request by gpanders**
Remove the "Compiled by" and "Compiled @" lines from :version. Also remove the compiled CFLAGS on non debug builds.

I don't know CMake very well, so I'm pretty confident this could be done better.

### CODE REVIEW

**Comment by justinmk** [comment]
if we need this we should probably guard it with a semantically-named flag like `NVIM_REPRODUCIBLE_BUILD` , rather than a narrow name `NVIM_INCLUDE_VERSION_CFLAGS`.

however I thought debian was already doing reproducible builds somehow, @jamessan would this be useful there?

**Comment by jamessan** [comment]
We are doing reproducible build testing.  Neovim currently [fails that](https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/neovim.html), but not due to the user/hostname issue.  Those are [explicitly set](https://salsa.debian.org/vim-team/neovim/-/blob/d5a8e5982b8ec9789f6a828d1fd4acbb439e02df/debian/rules#L30-31) when we build.

The bigger issue is that, at least for the Debian build, the build path shows up in the compiler flags, ironically enough, due to a flag which tries to help eliminate build path differences. :)

```
% nvim --version
NVIM v0.4.4
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -g -O2 -fdebug-prefix-map=/build/neovim-eK2ojg/neovim-0.4.4=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=1 -DDISABLE_LOG -Wdate-time -D_FORTIFY_SOURCE=1 -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim-eK2ojg/neovim-0.4.4/build/config -I/build/neovim-eK2ojg/neovim-0.4.4/src -I/usr/include -I/usr/include/lua5.1 -I/build/neovim-eK2ojg/neovim-0.4.4/build/src/nvim/auto -I/build/neovim-eK2ojg/neovim-0.4.4/build/include
Compiled by team+vim@tracker.debian.org

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

I don't think there's a need to remove the user/hostname bits.  I'm a bit torn on the compiler flags.  They do provide some use, but that use is fairly minimal and it is a more relevant issue for reproducibility.  As described on the reproducible builds results for neovim, there are efforts to try and address this specific issue more broadly, but it doesn't appear that has happened yet.

File: src/nvim/version.c [hunk 1]
```diff
@@ -2143,24 +2146,9 @@ void list_version(void)
   MSG(longVersion);
   MSG(version_buildtype);
   list_lua_version();
+#if NVIM_INCLUDE_VERSION_CFLAGS == 1
```
Comment (justinmk) [inline src/nvim/version.c:2149]: Maybe we can side-step the issue by just omitting it in Debug builds. This allows "bake-time" to see if we ever care about its absence. Then we can decide whether to remove it entirely.

```suggestion
  // Reproducible builds: omit compile info in Release builds.
#ifndef NDEBUG
```

**Comment by gpanders** [comment]
> Neovim currently fails that, but not due to the user/hostname issue... I don't think there's a need to remove the user/hostname bits.

(*With the huge caveat that I am mostly a novice in this area*) Unless I'm misunderstanding the concept of reproducible builds I don't think I agree. As I understand it, the idea is essentially that you and I can both start with the same source code (including library versions and OS, etc.) and produce the same binary, but this is impossible with the user/hostname bits (of course we can both manually set USERNAME and HOSTNAME as the Debian script does, and maybe this is sufficient, although it feels a bit loophole-y to me).

**Comment by justinmk** [comment]
> of course we can both manually set USERNAME and HOSTNAME as the Debian script does, and maybe this is sufficient, although it feels a bit loophole-y

It's not really a loophole, it is technically equivalent to setting a hypthoetical flag such as NVIM_INCLUDE_VERSION_CFLAGS.

Prefer gating it on NDEBUG anyways, with an eye towards total removal in the future.

**Comment by jamessan** [comment]
> but this is impossible with the user/hostname bits (of course we can both manually set USERNAME and HOSTNAME as the Debian script does, and maybe this is sufficient, although it feels a bit loophole-y to me).

Reproducibility requires building in the same way.  The intent is to verify that the same source, built in the same way but in a reasonably different environment, produces the same results.  This is no different than requiring the same `CFLAGS` between builds.  If you used `-O1` and I used `-O2`, then there's no way the resulting binaries would match.

This the first item mentioned in ["How?" to get reproducible builds](https://reproducible-builds.org/#how):

> Second, the set of tools used to perform the build and more generally the build environment should either be recorded or pre-defined.

Setting the host and username is pre-defining that portion of the environment.

**Comment by gpanders** [comment]
> > but this is impossible with the user/hostname bits (of course we can both manually set USERNAME and HOSTNAME as the Debian script does, and maybe this is sufficient, although it feels a bit loophole-y to me).
> 
> Reproducibility requires building in the same way. The intent is to verify that the same source, built in the same way but in a reasonably different environment, produces the same results. This is no different than requiring the same `CFLAGS` between builds. If you used `-O1` and I used `-O2`, then there's no way the resulting binaries would match.
> 
> This the first item mentioned in ["How?" to get reproducible builds](https://reproducible-builds.org/#how):
> 
> > Second, the set of tools used to perform the build and more generally the build environment should either be recorded or pre-defined.
> 
> Setting the host and username is pre-defining that portion of the environment.

That makes sense, thanks for the explanation.

**Pull request merged**